From b6b8df25d1b5d0c698bb9aaa0104637d55840903 Mon Sep 17 00:00:00 2001
From: Juha Kallioinen <juha.kallioinen@jolla.com>
Date: Tue, 12 Nov 2013 00:59:25 +0000
Subject: [PATCH 03/40] Fix memory leak when running plugins

---
 src/pk-transaction.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/pk-transaction.c b/src/pk-transaction.c
index c7e0350c6..32d58e629 100644
--- a/src/pk-transaction.c
+++ b/src/pk-transaction.c
@@ -996,6 +996,8 @@ pk_transaction_plugin_phase (PkTransaction *transaction,
 
 		/* quit the transaction if any of the plugins fail */
 		exit_code = pk_backend_job_get_exit_code (job);
+		g_object_unref(job);
+		job = NULL;
 		if (exit_code != PK_EXIT_ENUM_UNKNOWN &&
 		    exit_code != PK_EXIT_ENUM_SUCCESS) {
 			pk_backend_job_set_exit_code (transaction->priv->job, exit_code);
-- 
2.13.6

