From 673791c1a03c228df731ca010cf91c94cd8969bc Mon Sep 17 00:00:00 2001
From: Zhang Qiang <qiang.z.zhang@intel.com>
Date: Mon, 18 Apr 2011 16:07:29 +0800
Subject: [PATCH 05/40] zypp:support no_proxy for pkcon and zypp backend

---
 client/pk-console.c | 10 +++++++++-
 1 file changed, 9 insertions(+), 1 deletion(-)

diff --git a/client/pk-console.c b/client/pk-console.c
index 664fbecef..ef9b2ab3a 100644
--- a/client/pk-console.c
+++ b/client/pk-console.c
@@ -1346,7 +1346,11 @@ main (int argc, char *argv[])
 	gchar *summary = NULL;
 	const gchar *mode;
 	const gchar *http_proxy;
+	const gchar *https_proxy;
 	const gchar *ftp_proxy;
+	const gchar *socks_proxy;
+	const gchar *no_proxy;
+	const gchar *pac;
 	const gchar *value = NULL;
 	const gchar *details = NULL;
 	const gchar *parameter = NULL;
@@ -1467,10 +1471,14 @@ main (int argc, char *argv[])
 
 	/* set the proxy */
 	http_proxy = g_getenv ("http_proxy");
+	https_proxy = g_getenv ("https_proxy");
+	no_proxy = g_getenv ("no_proxy");
+	socks_proxy = g_getenv ("socks_proxy");
 	ftp_proxy = g_getenv ("ftp_proxy");
+	pac = g_getenv ("pac");
 	if (http_proxy != NULL ||
 	    ftp_proxy != NULL) {
-		ret = pk_control_set_proxy (control, http_proxy, ftp_proxy, NULL, &error_local);
+		ret = pk_control_set_proxy2 (control, http_proxy, https_proxy, ftp_proxy, socks_proxy, no_proxy, pac, NULL, &error_local);
 		if (!ret) {
 			/* TRANSLATORS: The user specified an incorrect filter */
 			error = g_error_new (1, 0, "%s: %s", _("The proxy could not be set"), error_local->message);
-- 
2.13.6

