From a899de99d9ff62f5df7fbca61ff9de8fd7ebe6ee Mon, 10 Oct 2016 15:48:17 +0200
From: Zhang Qiang <qiang.z.zhang@intel.com>
Date: Mon, 18 Apr 2011 16:07:29 +0800
Subject: [PATCH] zypp:support no_proxy for pkcon and zypp backend


diff --git a/client/pk-console.c b/client/pk-console.c
index 664fbec..ef9b2ab 100644
--- a/client/pk-console.c
+++ b/client/pk-console.c
@@ -1346,7 +1346,11 @@
 	gchar *summary = NULL;
 	const gchar *mode;
 	const gchar *http_proxy;
+	const gchar *https_proxy;
 	const gchar *ftp_proxy;
+	const gchar *socks_proxy;
+	const gchar *no_proxy;
+	const gchar *pac;
 	const gchar *value = NULL;
 	const gchar *details = NULL;
 	const gchar *parameter = NULL;
@@ -1467,10 +1471,14 @@
 
 	/* set the proxy */
 	http_proxy = g_getenv ("http_proxy");
+	https_proxy = g_getenv ("https_proxy");
+	no_proxy = g_getenv ("no_proxy");
+	socks_proxy = g_getenv ("socks_proxy");
 	ftp_proxy = g_getenv ("ftp_proxy");
+	pac = g_getenv ("pac");
 	if (http_proxy != NULL ||
 	    ftp_proxy != NULL) {
-		ret = pk_control_set_proxy (control, http_proxy, ftp_proxy, NULL, &error_local);
+		ret = pk_control_set_proxy2 (control, http_proxy, https_proxy, ftp_proxy, socks_proxy, no_proxy, pac, NULL, &error_local);
 		if (!ret) {
 			/* TRANSLATORS: The user specified an incorrect filter */
 			error = g_error_new (1, 0, "%s: %s", _("The proxy could not be set"), error_local->message);
