From 58f338fe70a718442d7abe23fe09f08b840283a9 Mon Sep 17 00:00:00 2001
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 12 Nov 2013 14:47:26 +0100
Subject: [PATCH 21/38] zypp: Remove old leftover PK_TMP_DIR on localinstall

---
 backends/zypp/pk-backend-zypp.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 24ecdf94e..60017d211 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -2640,6 +2640,15 @@ backend_install_files_thread (PkBackendJob *job, GVariant *params, gpointer user
 		}
 	}
 
+	// Remove a possibly leftover temporary repository
+	try {
+		RepoInfo oldRepo = manager.getRepositoryInfo("PK_TMP_DIR");
+		manager.removeRepository(oldRepo);
+		PK_ZYPP_LOG("Removed old temporary install repository");
+	} catch (const Exception &ex) {
+		// Repo probably doesn't exist yet - that's GOOD :)
+	}
+
 	// create a plaindir-repo and cache it
 	RepoInfo tmpRepo;
 
-- 
2.13.6

