From 784e061fa324793fe278c8f3f0103ca7454a7faf Mon Sep 17 00:00:00 2001
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 12 Nov 2013 14:47:26 +0100
Subject: [PATCH 23/40] zypp: Remove old leftover PK_TMP_DIR on localinstall

---
 backends/zypp/pk-backend-zypp.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index f844fb32f..d89a4126d 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -2610,6 +2610,15 @@ backend_install_files_thread (PkBackendJob *job, GVariant *params, gpointer user
 		}
 	}
 
+	// Remove a possibly leftover temporary repository
+	try {
+		RepoInfo oldRepo = manager.getRepositoryInfo("PK_TMP_DIR");
+		manager.removeRepository(oldRepo);
+		PK_ZYPP_LOG("Removed old temporary install repository");
+	} catch (const Exception &ex) {
+		// Repo probably doesn't exist yet - that's GOOD :)
+	}
+
 	// create a plaindir-repo and cache it
 	RepoInfo tmpRepo;
 
-- 
2.13.6

