From b910af931b0b92126088e297febdbb1d939c889b Mon Sep 17 00:00:00 2001
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 11 Feb 2014 11:13:54 +0100
Subject: [PATCH 23/38] Validate RPM arch before local-install. Fixes JB#15666

---
 backends/zypp/pk-backend-zypp.cpp | 9 +++++++++
 1 file changed, 9 insertions(+)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index a2e876cf0..31a5666e1 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -2623,6 +2623,15 @@ backend_install_files_thread (PkBackendJob *job, GVariant *params, gpointer user
 			return;
 		}
 
+		// check if the file has an acceptable architecture
+		if (! rpmHeader->tag_arch().compatibleWith (ZConfig::instance().systemArchitecture ())) {
+			zypp_backend_finished_error (
+				job, PK_ERROR_ENUM_LOCAL_INSTALL_FAILED,
+				"%s has wrong architecture: %s", full_paths[i],
+				rpmHeader->tag_arch ().asString (). c_str());
+			return;
+		}
+
 		// copy the rpm into tmpdir
 		string tempDest = tmpDir.path ().asString () + "/" + rpmHeader->tag_name () + ".rpm";
 		if (filesystem::copy (full_paths[i], tempDest) != 0) {
-- 
2.13.6

