From b6172874ab2db87281dc725581de3347b7c5f0a4 Mon Sep 17 00:00:00 2001
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 11 Mar 2014 13:20:52 +0100
Subject: [PATCH 24/38] Clear metadata when disabling repositories. JB#15874

---
 backends/zypp/pk-backend-zypp.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index e9012bc10..2b60a3529 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -3361,6 +3361,7 @@ pk_backend_repo_enable (PkBackend *backend, PkBackendJob *job, const gchar *rid,
 		repo.setEnabled (enabled);
 		manager.modifyRepository (rid, repo);
 		if (!enabled) {
+			manager.cleanMetadata (repo);
 			Repository repository = sat::Pool::instance ().reposFind (repo.alias ());
 			repository.eraseFromPool ();
 		}
-- 
2.13.6

