From 9559e7daf7db28927d3ad77417056a55c8975438 Mon Sep 17 00:00:00 2001
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 3 Jun 2014 15:48:06 +0200
Subject: [PATCH 49/65] Update whitelisted applications' binary names

---
 src/pk-transaction.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/src/pk-transaction.c b/src/pk-transaction.c
index ec154c745..1898fa28b 100644
--- a/src/pk-transaction.c
+++ b/src/pk-transaction.c
@@ -2847,10 +2847,10 @@ pk_transaction_builtin_policy_allow (guint pid)
 		"/usr/bin/store-client",
 
 		// For devel dist upgrading
-		"/usr/bin/devel-dist-upgrade",
+		"/usr/bin/rnd-dist-upgrade",
 
 		// For deploying RPMs from the SDK
-		"/usr/bin/devel-deploy-rpm",
+		"/usr/bin/sdk-deploy-rpm",
 
 		NULL,
 	};
-- 
2.13.5

