From 0a4cf815f0f4e6961ea1130a20a0f536eceb4d55 Mon, 10 Oct 2016 15:50:52 +0200
From: Thomas Perl <thomas.perl@jollamobile.com>
Date: Mon, 19 Aug 2013 12:09:42 +0200
Subject: [PATCH] zypp Backend: Close PackageKit job when upgrade fails


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index bdff077..b7eb695 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -3813,7 +3813,7 @@
 		ResPool pool = zypp_build_pool (zypp, TRUE);
 
 		if (!zypp_perform_execution (job, zypp, UPGRADE, TRUE, transaction_flags)) {
-			return;
+			MIL << "Upgrade execution failed" << std::endl;
 		}
 	} catch (const Exception &ex) {
 		zypp_backend_finished_error (job,
