From 76ab96347cf69fda8ae92a4a8c812cb9ca34873d Mon, 10 Oct 2016 15:52:00 +0200
From: Thomas Perl <thomas.perl@jollamobile.com>
Date: Fri, 4 Oct 2013 12:56:14 +0200
Subject: [PATCH] [backends] zypp: Fix double free in libzypp (TmpLineWriter takes ownership)


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index be23058..ea2f45c 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -226,7 +226,6 @@
 public:
 	SystemdLogWriter()
 		: zypp::log::LineWriter()
-		  , m_writer(this)
 	{
 	}
 
@@ -248,9 +247,6 @@
 			}
 		}
 	}
-
-private:
-	zypp::base::LogControl::TmpLineWriter m_writer;
 };
 
 
@@ -662,7 +658,7 @@
 	std::vector<std::string> signatures;
 	EventDirector eventDirector;
 	PkBackendJob *currentJob;
-	SystemdLogWriter logWriter;
+	zypp::base::LogControl::TmpLineWriter *tmpLineWriter;
 	
 	pthread_mutex_t zypp_mutex;
 	ExecCounters exec;
@@ -1903,6 +1899,8 @@
 	priv->currentJob = 0;
 	priv->zypp_mutex = PTHREAD_MUTEX_INITIALIZER;
 	priv->exec = ExecCounters();
+	// tmpLineWriter takes ownership of new SystemdLogWriter
+	priv->tmpLineWriter = new zypp::base::LogControl::TmpLineWriter(new SystemdLogWriter);
 
 	g_debug ("zypp_backend_initialize");
 	//_updating_self = FALSE;
@@ -1917,6 +1915,10 @@
 {
 	g_debug ("zypp_backend_destroy");
 
+	if (priv->tmpLineWriter) {
+		delete priv->tmpLineWriter;
+	}
+
 	g_free (_repoName);
 	delete priv;
 }
