From 262c5e56b726eac0477f488fc741d8304530f541 Mon, 10 Oct 2016 15:54:18 +0200
From: Juha Kallioinen <juha.kallioinen@jolla.com>
Date: Mon, 28 Oct 2013 17:49:39 +0000
Subject: [PATCH] [backends] Call db_recovery in case an rpm db corruption is noticed


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index a23690e..d847405 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -767,12 +767,25 @@
 {
 	PK_ZYPP_LOG("Exception while initializing target (RPM DB broken?): %s", e.asUserString().c_str());
 
-	// TODO: Repair RPM db, then re-initializeTarget() and if that works, return true
-	//filesystem::Pathname pathname("/");
-	//zypp->initializeTarget (pathname);
+	if (system("/usr/bin/db_recover -h /var/lib/rpm") == EXIT_SUCCESS) {
+		PK_ZYPP_LOG("RPM DB recovery successful.");
 
-	// Right now, we can't repair the database, so fail instead
-	return false;
+		try {
+			filesystem::Pathname pathname("/");
+			zypp->initializeTarget (pathname);
+			return true;
+		}
+		catch(const Exception &e) {
+			PK_ZYPP_LOG("RPM DB reinitialization failed.");
+			return false;
+		}
+	}
+	else {
+		PK_ZYPP_LOG("RPM DB recovery failed.");
+		// TODO should we fork and do rm -f /var/lib/rpm/__db*; rpm --rebuilddb here?
+		return false;
+
+	}
 }
 
 /**
