From 6e15932c47b52b8e565939743158b8638f41ed36 Mon, 10 Oct 2016 15:55:29 +0200
From: Juha Kallioinen <juha.kallioinen@jolla.com>
Date: Tue, 12 Nov 2013 00:59:25 +0000
Subject: [PATCH] [pk-transaction] Fix memory leak when running plugins


diff --git a/src/pk-transaction.c b/src/pk-transaction.c
index c7e0350..32d58e6 100644
--- a/src/pk-transaction.c
+++ b/src/pk-transaction.c
@@ -996,6 +996,8 @@
 
 		/* quit the transaction if any of the plugins fail */
 		exit_code = pk_backend_job_get_exit_code (job);
+		g_object_unref(job);
+		job = NULL;
 		if (exit_code != PK_EXIT_ENUM_UNKNOWN &&
 		    exit_code != PK_EXIT_ENUM_SUCCESS) {
 			pk_backend_job_set_exit_code (transaction->priv->job, exit_code);
