From 0925c4c19f8a9fa5d4fe4f0ab65d48e222d77b73 Mon, 10 Oct 2016 15:55:42 +0200
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 12 Nov 2013 14:47:26 +0100
Subject: [PATCH] [backends] zypp: Remove old leftover PK_TMP_DIR on localinstall


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index bec3e3e..7694a09 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -2608,6 +2608,15 @@
 		}
 	}
 
+	// Remove a possibly leftover temporary repository
+	try {
+		RepoInfo oldRepo = manager.getRepositoryInfo("PK_TMP_DIR");
+		manager.removeRepository(oldRepo);
+		PK_ZYPP_LOG("Removed old temporary install repository");
+	} catch (const Exception &ex) {
+		// Repo probably doesn't exist yet - that's GOOD :)
+	}
+
 	// create a plaindir-repo and cache it
 	RepoInfo tmpRepo;
 
