From 43e63f676eb64029edbeb08a19eddf1b001a1f22 Mon, 10 Oct 2016 16:03:48 +0200
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 11 Mar 2014 13:20:52 +0100
Subject: [PATCH] [zypp] Clear metadata when disabling repositories. JB#15874


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index e69dcee..aa443bf 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -3341,6 +3341,7 @@
 		repo.setEnabled (enabled);
 		manager.modifyRepository (rid, repo);
 		if (!enabled) {
+			manager.cleanMetadata (repo);
 			Repository repository = sat::Pool::instance ().reposFind (repo.alias ());
 			repository.eraseFromPool ();
 		}
