From db1558438dd6e5fbe31517f064acfe5269e9591a Mon, 10 Oct 2016 16:04:18 +0200
From: Thomas Perl <thomas.perl@jolla.com>
Date: Wed, 2 Apr 2014 15:44:12 +0200
Subject: [PATCH] [zypp] Handle and log C++ exceptions in all threads


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index aa443bf..6ff0eab 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -210,6 +210,62 @@
 	return package_id;
 }
 
+class ZyppBackendThreadWrapperData {
+public:
+	ZyppBackendThreadWrapperData(PkBackendJobThreadFunc func,
+			gpointer user_data, GDestroyNotify destroy_func)
+		: func(func)
+		, user_data(user_data)
+		, destroy_func(destroy_func)
+	{
+	}
+
+	~ZyppBackendThreadWrapperData()
+	{
+	}
+
+	PkBackendJobThreadFunc func;
+	gpointer user_data;
+	GDestroyNotify destroy_func;
+};
+
+static void
+zypp_backend_job_thread_wrapper (PkBackendJob *job, GVariant *params,
+		gpointer user_data)
+{
+	ZyppBackendThreadWrapperData *data = static_cast<ZyppBackendThreadWrapperData *>(user_data);
+
+	try {
+		// Call real thread function
+		data->func(job, params, data->user_data);
+	} catch (const Exception &ex) {
+		PK_ZYPP_LOG("C++ exception in zypp backend job: %s",
+				ex.asUserString().c_str());
+		pk_backend_job_error_code (job, PK_ERROR_ENUM_INTERNAL_ERROR,
+				ex.asUserString().c_str());
+	}
+
+	delete data;
+}
+
+/**
+ * Wrapper function that takes care of catching C++ exceptions for us if
+ * they are not caught somewhere more specific in the call stack. A C++
+ * exception thrown over the PackageKit backend API would lead to an
+ * unconditional abort, possibly corrupting the rpmdb, so we catch it here
+ * and make sure we propagate the exception message to the job as error,
+ * log the error and avoid general crash'n'burn behavior.
+ **/
+static gboolean
+zypp_backend_job_thread_create (PkBackendJob *job, PkBackendJobThreadFunc func,
+		gpointer user_data, GDestroyNotify destroy_func)
+{
+	ZyppBackendThreadWrapperData *data = new ZyppBackendThreadWrapperData(func,
+			user_data, destroy_func);
+	return pk_backend_job_thread_create (job, zypp_backend_job_thread_wrapper,
+			data, destroy_func);
+}
+
 static int64_t
 get_free_disk_space(const char *path)
 {
@@ -2100,7 +2156,7 @@
 void
 pk_backend_get_requires(PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **package_ids, gboolean recursive)
 {
-	pk_backend_job_thread_create (job, backend_get_requires_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_requires_thread, NULL, NULL);
 }
 
 /**
@@ -2293,7 +2349,7 @@
 void
 pk_backend_get_depends (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **package_ids, gboolean recursive)
 {
-	pk_backend_job_thread_create (job, backend_get_depends_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_depends_thread, NULL, NULL);
 }
 
 static void
@@ -2365,7 +2421,7 @@
 void
 pk_backend_get_details (PkBackend *backend, PkBackendJob *job, gchar **package_ids)
 {
-	pk_backend_job_thread_create (job, backend_get_details_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_details_thread, NULL, NULL);
 }
 
 static void
@@ -2422,7 +2478,7 @@
 void
 pk_backend_get_distro_upgrades (PkBackend *backend, PkBackendJob *job)
 {
-	pk_backend_job_thread_create (job, backend_get_distro_upgrades_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_distro_upgrades_thread, NULL, NULL);
 }
 
 static void
@@ -2446,7 +2502,7 @@
 void
 pk_backend_refresh_cache (PkBackend *backend, PkBackendJob *job, gboolean force)
 {
-	pk_backend_job_thread_create (job, backend_refresh_cache_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_refresh_cache_thread, NULL, NULL);
 }
 
 /* If a critical self update (see qualifying steps below) is available then only show/install that update first.
@@ -2554,7 +2610,7 @@
 void
 pk_backend_get_updates (PkBackend *backend, PkBackendJob *job, PkBitfield filters)
 {
-	pk_backend_job_thread_create (job, backend_get_updates_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_updates_thread, NULL, NULL);
 }
 
 static void
@@ -2683,7 +2739,7 @@
 void
 pk_backend_install_files (PkBackend *backend, PkBackendJob *job, PkBitfield transaction_flags, gchar **full_paths)
 {
-	pk_backend_job_thread_create (job, backend_install_files_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_install_files_thread, NULL, NULL);
 }
 
 static void
@@ -2779,7 +2835,7 @@
 void
 pk_backend_get_update_detail (PkBackend *backend, PkBackendJob *job, gchar **package_ids)
 {
-	pk_backend_job_thread_create (job, backend_get_update_detail_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_update_detail_thread, NULL, NULL);
 }
 
 static void
@@ -2871,7 +2927,7 @@
 {
 	// For now, don't let the user cancel the install once it's started
 	pk_backend_job_set_allow_cancel (job, FALSE);
-	pk_backend_job_thread_create (job, backend_install_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_install_packages_thread, NULL, NULL);
 }
 
 
@@ -2897,7 +2953,7 @@
 void
 pk_backend_install_signature (PkBackend *backend, PkBackendJob *job, PkSigTypeEnum type, const gchar *key_id, const gchar *package_id)
 {
-	pk_backend_job_thread_create (job, backend_install_signature_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_install_signature_thread, NULL, NULL);
 }
 
 static void
@@ -2990,7 +3046,7 @@
 pk_backend_remove_packages (PkBackend *backend, PkBackendJob *job, PkBitfield transaction_flags,
 			    gchar **package_ids, gboolean allow_deps, gboolean autoremove)
 {
-	pk_backend_job_thread_create (job, backend_remove_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_remove_packages_thread, NULL, NULL);
 }
 
 static void
@@ -3084,7 +3140,7 @@
 void
 pk_backend_resolve (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **package_ids)
 {
-	pk_backend_job_thread_create (job, backend_resolve_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_resolve_thread, NULL, NULL);
 }
 
 static void
@@ -3180,7 +3236,7 @@
 void
 pk_backend_search_names (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **values)
 {
-	pk_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
 }
 
 /**
@@ -3189,7 +3245,7 @@
 void
 pk_backend_search_details (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **values)
 {
-	pk_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
 }
 
 static void
@@ -3252,7 +3308,7 @@
 void
 pk_backend_search_groups (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **values)
 {
-	pk_backend_job_thread_create (job, backend_search_group_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_search_group_thread, NULL, NULL);
 }
 
 /**
@@ -3261,7 +3317,7 @@
 void
 pk_backend_search_files (PkBackend *backend, PkBackendJob *job, PkBitfield filters, gchar **values)
 {
-	pk_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_find_packages_thread, NULL, NULL);
 }
 
 /**
@@ -3419,7 +3475,7 @@
 void
 pk_backend_get_files(PkBackend *backend, PkBackendJob *job, gchar **package_ids)
 {
-	pk_backend_job_thread_create (job, backend_get_files_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_files_thread, NULL, NULL);
 }
 
 static void
@@ -3462,7 +3518,7 @@
 void
 pk_backend_get_packages (PkBackend *backend, PkBackendJob *job, PkBitfield filter)
 {
-	pk_backend_job_thread_create (job, backend_get_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_get_packages_thread, NULL, NULL);
 }
 
 static void
@@ -3526,7 +3582,7 @@
 void
 pk_backend_update_packages (PkBackend *backend, PkBackendJob *job, PkBitfield transaction_flags, gchar **package_ids)
 {
-	pk_backend_job_thread_create (job, backend_update_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_update_packages_thread, NULL, NULL);
 }
 
 static void
@@ -3666,7 +3722,7 @@
 void
 pk_backend_repo_set_data (PkBackend *backend, PkBackendJob *job, const gchar *repo_id, const gchar *parameter, const gchar *value)
 {
-	pk_backend_job_thread_create (job, backend_repo_set_data_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_repo_set_data_thread, NULL, NULL);
 }
 
 /**
@@ -3824,7 +3880,7 @@
 void
 pk_backend_what_provides (PkBackend *backend, PkBackendJob *job, PkBitfield filters, PkProvidesEnum provide, gchar **values)
 {
-	pk_backend_job_thread_create (job, backend_what_provides_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_what_provides_thread, NULL, NULL);
 }
 
 gchar **
@@ -3920,7 +3976,7 @@
 void
 pk_backend_download_packages (PkBackend *backend, PkBackendJob *job, gchar **package_ids, const gchar *directory)
 {
-	pk_backend_job_thread_create (job, backend_download_packages_thread, NULL, NULL);
+	zypp_backend_job_thread_create (job, backend_download_packages_thread, NULL, NULL);
 }
 
 /**
@@ -4037,7 +4093,7 @@
 pk_backend_upgrade_system (PkBackend *backend, PkBackendJob *job,
 	const gchar *distro_id, PkUpgradeKindEnum upgrade_kind)
 {
-	pk_backend_job_thread_create (job, backend_upgrade_system_thread,
+	zypp_backend_job_thread_create (job, backend_upgrade_system_thread,
 			new DistUpgrade(distro_id, upgrade_kind), NULL);
 }
 
