From d8abe9f41d91365d2fad841ba4d6080771617484 Mon, 10 Oct 2016 16:10:04 +0200
From: Juha Kallioinen <juha.kallioinen@jolla.com>
Date: Fri, 18 Apr 2014 22:40:31 +0000
Subject: [PATCH] [zypp] accept i486 packages - JB#18558


Signed-off-by: Juha Kallioinen <juha.kallioinen@jolla.com>

diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 83d0d63..fc1de52 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -1339,8 +1339,8 @@
 static gboolean
 system_and_package_are_x86 (Arch arch)
 {
-	// i586, i686, ... all should be considered the same arch for our comparison
-	return ( arch == Arch_i586 && ZConfig::defaultSystemArchitecture() == Arch_i686 );
+	// i486, i586, i686, ... all should be considered the same arch for our comparison
+	return ((arch == Arch_i486 || arch == Arch_i586 ) && ZConfig::defaultSystemArchitecture() == Arch_i686 );
 }
 
 static gboolean
