From fe7cc4887cc24318489b6dd4bddc139b8baaf648 Mon, 10 Oct 2016 16:10:34 +0200
From: Thomas Perl <thomas.perl@jolla.com>
Date: Thu, 15 May 2014 11:17:31 +0200
Subject: [PATCH] [zypp] Allow searching for capabilities


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 36e71e6..4fa7522 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -3279,7 +3279,9 @@
 	vector<sat::Solvable> v;
 
 	PoolQuery q;
-	q.addString( search ); // may be called multiple times (OR'ed)
+	Capability cap (search);
+	std::string name = cap.detail().name().asString();
+	q.addString( name ); // may be called multiple times (OR'ed)
 	q.setCaseSensitive( true );
 	q.setMatchSubstring();
 
@@ -3290,6 +3292,14 @@
 		q.addKind( ResKind::pattern );
 		q.addKind( ResKind::srcpackage );
 		q.addAttribute( sat::SolvAttr::name );
+
+		// Search also Provides: to allow querying for "debuginfo(build-id)=<hash>"
+		q.addDependency( sat::SolvAttr::provides,
+				 name,
+				 cap.detail().op(),
+				 cap.detail().ed(),
+				 Arch(cap.detail().arch()) );
+
 		// Note: The query result is NOT sorted packages first, then srcpackage.
 		// If that's necessary you need to sort the vector accordongly or use
 		// two separate queries.
