From e7bc7af340ad6d3bd62a64858f2b73ae7180c6f4 Mon, 10 Oct 2016 16:10:40 +0200
From: Thomas Perl <thomas.perl@jolla.com>
Date: Tue, 3 Jun 2014 15:48:06 +0200
Subject: [PATCH] [transaction] Update whitelisted applications' binary names


diff --git a/src/pk-transaction.c b/src/pk-transaction.c
index ec154c7..1898fa2 100644
--- a/src/pk-transaction.c
+++ b/src/pk-transaction.c
@@ -2847,10 +2847,10 @@
 		"/usr/bin/store-client",
 
 		// For devel dist upgrading
-		"/usr/bin/devel-dist-upgrade",
+		"/usr/bin/rnd-dist-upgrade",
 
 		// For deploying RPMs from the SDK
-		"/usr/bin/devel-deploy-rpm",
+		"/usr/bin/sdk-deploy-rpm",
 
 		NULL,
 	};
