From ca9be6f8213fd0d398a61dbcd105b5eb321c7b14 Mon, 10 Oct 2016 16:10:49 +0200
From: Sami Kananoja <sami.kananoja@jollamobile.com>
Date: Fri, 22 Aug 2014 14:15:30 +0300
Subject: [PATCH] [zypp] Take uninstalled due to upgrade packages into account in size calculations


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 4fa7522..265a398 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -1859,12 +1859,11 @@
 				priv->exec.total_downloads += 1;
 				// TODO: Only count download bytes if updates were not yet downloaded
 				total_download_bytes += it->resolvable()->downloadSize();
-			} else if (it->status().isToBeUninstalled() &&
-					!it->status().isToBeUninstalledDueToUpgrade()) {
-				if (!only_download) {
+			} else if (!only_download && it->status().isToBeUninstalled()) {
+				if (!it->status().isToBeUninstalledDueToUpgrade()) {
 					priv->exec.total_removals += 1;
-					total_remove_bytes += it->resolvable()->installSize();
 				}
+				total_remove_bytes += it->resolvable()->installSize();
 			}
 		}
 
