From 444b22f4b168db3741e1302fe9fe1993b353de2e Mon, 10 Oct 2016 16:10:58 +0200
From: Thomas Perl <m@thp.io>
Date: Fri, 16 Jan 2015 09:47:03 +0100
Subject: [PATCH] Fix SIGFPE when no packages are installed


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 4fa7522..475a18f 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -751,7 +751,10 @@
 			current = total;
 		}
 
-		int percentage = 100 * current / total;
+		int percentage = -1;
+		if (total != 0) {
+			percentage = 100 * current / total;
+		}
 
 		if (last_percentage > percentage) {
 			/**
