From 19c90b71f83df4377609ba823f5d948809b7d4b8 Mon, 10 Oct 2016 16:17:27 +0200
From: Martin Grimme <martin.grimme@gmail.com>
Date: Fri, 11 Sep 2015 13:05:28 +0200
Subject: [PATCH] [zypp] Fixed implementation of finding newest package while resolving. Fixes JB#32224


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 5d1b9b0..aaa4f61 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -3467,7 +3467,10 @@
 			
 			if (zypp_is_no_solvable(newest)) {
 				newest = *it;
-			} else if (it->edition() > newest.edition() || Arch::compare(it->arch(), newest.arch()) > 0) {
+			} else if (it->edition() > newest.edition()) {
+				newest = *it;
+			} else if (it->edition() == newest.edition()
+					&& Arch::compare(it->arch(), newest.arch()) > 0) {
 				newest = *it;
 			}
 			//MIL << "emit " << *it << endl;
