From b5713f57a22bcc9159a5b29c4f490b2d27479973 Thu, 13 Oct 2016 11:44:33 +0200
From: Martin Grimme <martin.grimme@gmail.com>
Date: Fri, 2 Oct 2015 09:26:48 +0200
Subject: [PATCH] [zypp] Revert dependency on libzypp version 14 as we cannot have it yet. Contributes to JB#31091


diff --git a/backends/zypp/pk-backend-zypp.cpp b/backends/zypp/pk-backend-zypp.cpp
index 20f9c38..def5d3e 100644
--- a/backends/zypp/pk-backend-zypp.cpp
+++ b/backends/zypp/pk-backend-zypp.cpp
@@ -2062,14 +2062,15 @@
 
 				Package::constPtr pkg = asKind<Package>(it->resolvable());
 				if (pkg) {
-					if (pkg->isCached()) {
-						total_cached_bytes += pkg->downloadSize();
-					} else {
+					// TODO: Enable once we have upgraded libzypp
+					//if (pkg->isCached()) {
+					//	total_cached_bytes += pkg->downloadSize();
+					//} else {
 						total_download_bytes += pkg->downloadSize();
 						if (pkg->downloadSize() > biggest_package_download) {
 							biggest_package_download = pkg->downloadSize();
 						}
-					}
+					//}
 				}
 			} else if (!only_download && it->status().isToBeUninstalled()) {
 				if (!it->status().isToBeUninstalledDueToUpgrade()) {
